module/VuFind/src/VuFind/Marc/Serialization/MarcInJson.php

```diff
@@ -0,0 +1,202 @@
+        if (false === ($f = fopen($file, 'rb'))) {
+            throw new \Exception("Cannot open file '$file' for reading");
+        }
+        $s = '';
+        do {
+            $s .= fgets($f, 10);
+        } while (strlen(ltrim($s)) < 5 && !feof($f));
+        fclose($f);
+
+        return self::canParseCollection($s);
+    }
+
+    /**
+     * Parse MARC collection from a string into an array
+     *
+     * @param string $collection MARC record collection in the format supported by
+     * the serialization class
+     *
+     * @throws \Exception
+     * @return array
+     */
+    public static function collectionFromString(string $collection): array
+    {
+        return json_decode($collection, true);
+    }
+
+    /**
+     * Parse MARC-in-JSON
+     *
+     * @param string $marc JSON
+     *
+     * @throws \Exception
+     * @return array
+     */
+    public static function fromString(string $marc): array
+    {
+        return json_decode($marc, true);
+    }
+
+    /**
+     * Convert record to ISO2709 string
+     *
+     * @param array $record Record data
+     *
+     * @return string
+     */
+    public static function toString(array $record): string
+    {
+        // We need to cast any subfield with '0' as key to an object to, otherwise
+        // it would be encoded as a simple array instead of an object:
+        foreach ($record['fields'] as &$fieldData) {
+            $field = current($fieldData);
+            if (!is_array($field)) {
+                continue;
+            }
+            foreach ($field['subfields'] as &$subfield) {
+                if (key($subfield) == 0) {
+                    $subfield = (object)$subfield;
+                }
+            }
+            unset($subfield);
+            $fieldData = [key($fieldData) => $field];
+        }
+        unset($fieldData);
+        return json_encode($record, JSON_UNESCAPED_UNICODE);
+    }
+
+    /**
+     * Open a collection file
+     *
+     * @param string $file File name
+     *
+     * @return void
+     */
+    public function openCollectionFile(string $file): void
+    {
+        $data = file_get_contents($file);
+        if (false === $data) {
+            throw new \Exception("Cannot open file '$file' for reading");
+        }
+        $this->collection = json_decode($data, true);
+        if (null === $this->collection) {
+            throw
+                new \Exception("File '$file' is invalid: " . json_last_error_msg());
+        }
+        $this->position = -1;
+    }
+
+    /**
+     * Rewind the collection file
+     *
+     * @return void;
+     */
+    public function rewind(): void
+    {
+        if (null === $this->collection) {
+            throw new \Exception('Collection not available');
+        }
+        $this->position = -1;
+    }
+
+    /**
+     * Get next record from the file or an empty string on EOF
+     *
+     * @return string
+     */
+    public function getNextRecord(): string
+    {
+        ++$this->position;
```

Comment:
This does not appear to be streaming support. Should we add a TODO or open a JIRA ticket? I realize that's probably a bigger project and not necessarily a priority right now... but we might want it in the future, so we shouldn't forget about it.

Reply:
I added proper streaming support. :)